aggregator/src/aggregator/collection_job_driver.rs

```diff
@@ -302,18 +309,31 @@ where
                         &batch_aggregations,
                     );
 
-                    try_join!(
-                        try_join_all(
-                            batch_aggregations
-                                .iter()
-                                .map(|ba| tx.update_batch_aggregation(ba))
-                        ),
-                        try_join_all(
-                            empty_batch_aggregations
-                                .iter()
-                                .map(|ba| tx.put_batch_aggregation(ba))
-                        ),
-                    )?;
+                    // Rather than awaiting all the futures at once, which can cause heap growth
+                    // proportional to the number of batch aggregations to write, put them into an
+                    // unordered, buffered stream so that they get run in groups of bounded size.
+                    // #3857
+                    {
+                        // This scope is needed so the borrows of batch_aggregations and
+                        // empty_batch_aggregations get dropped
+                        let mut futures = Vec::with_capacity(
+                            batch_aggregations.len() + empty_batch_aggregations.len(),
+                        );
+                        for ba in batch_aggregations.iter() {
+                            futures.push(tx.update_batch_aggregation(ba).boxed());
+                        }
+                        for ba in empty_batch_aggregations.iter() {
+                            futures.push(tx.put_batch_aggregation(ba).boxed());
+                        }
+
+                        // Drain the stream. This will short circuit if any of the futures return
+                        // an error.
+                        let mut stream =
+                            futures::stream::iter(futures).buffer_unordered(max_future_concurrency);
+                        while let Some(r) = stream.next().await {
+                            r?;
+                        }
```

Comment:
This loop looks like we're running the futures one at a time. But from [the docs](https://docs.rs/futures/0.3.31/futures/stream/trait.StreamExt.html#method.buffer_unordered):

> If this stream’s item can be converted into a future, then this adaptor will buffer up to n futures and then return the outputs in the order in which they complete.

So my understanding is that each group of `max_future_concurrency` futures will get `await`ed for us with the desired concurrency by `buffer_unordered` and this loop just serializes checking for errors.